<h1 align='center'> 파이썬과학프로그래밍기초 실습 2</h1>

<p style="text-align: right"> 2022.3.17</p>

<div class="alert alert-block alert-info">
    
- 파일 이름에서 20185124을 자신의 학번으로, name을 자신의 이름으로 수정하세요.

- 다음 줄에 자신의 이름, 학번, 학과(전공)을 적으세요.

* 이름:   나영훈          학번:   20185124         학과(전공):  빅데이터 전공
    
</div>

- JupyterLab 문서의 최신 버전은 [JupyterLab Documentation](https://jupyterlab.readthedocs.io/en/stable/index.html#/)을  참고하라

- Markdown은 [Markdown Guide](https://www.markdownguide.org/)를 참고하라.
- [Markdown Cheat Sheet](https://www.markdownguide.org/cheat-sheet/)

* 제출 마감: 3월 23일 (수) 오후 6:00까지 최종본 SmartLEAD에 제출

In [ ]:
import sys
sys.version

## 객체들의 비교 연산

### 예제 1. 
다음 연산 결과가 나온 이유를 설명하라.

In [1]:
45 > 5

True

**설명**: 45라는 값이 5보다 크기 때문

In [2]:
3 > 3+ 3j

TypeError: '>' not supported between instances of 'int' and 'complex'

**설명**: 복소수 연산 (complex 자료형)은 비교가 불가능.

In [3]:
'a' > 'b'

False

In [7]:
ord('a'), ord('b')

(97, 98)

**설명**: 유니코드 상에서 'a'는 97, 'b'는 98이기 때문

In [14]:
'가' > 'abc'

True

In [13]:
ord('가'), ord('a')

(44032, 97)

**설명**: 가와 a는 유니코드 상에서 44032, 97이다. 따라서 `가` 가 더 크기 때문에 True 를 반환한다.

In [9]:
[1, 5, 'tabc'] > [1, 5, 'str']

True

In [12]:
ord('t'), ord('s')

(116, 115)

**설명**: 앞의 1이 같으므로 뒤인 5로, 5도 같으므로 뒤로 넘김. 뒤에서는 
't'와, 's' 의 값이 차이가 나기 떄문에 T/F중 하나를 반환해야함. 유니코드 상에서 t와 s는 각각 116, 115이므로 True를 반환.

In [15]:
(2, 2) > (1, 1, 5)

True

**설명**: 처음 값 2와 1을 비교했을 때 왼쪽이 더 크기 때문. 연산의 결과가 나왔으므로 뒤는 비교하지 않는다.

## 정수의 비트 단위 연산

### 예제 2. 
다음 연산 과정을 설명하라

In [16]:
5&6, 5|6, 5^6 

(4, 7, 3)

In [25]:
bin(5)[2:], bin(6)[2:]

#101
#110 

#100 (and) -> 4
#111 (or)  -> 7
#011 (xor) -> 3

('101', '110')

**설명**: bin()을 통해 5와 6을 2진수로 변환후 And, or, xor 연산을 하였습니다.

- `&` : and 연산자로 피연산자값 값이 모두 1이여야 1이 됨, 1이 하나만 있다면 0을 출력. 따라서 $ 100_{(2)} \rightarrow 4_{(10)} $  을 출력
- `|` : or 연산자로 피연산자값이 하나만 1이여도 1을 출력, 모두 0일 때만 0을 출력.  따라서 $ 111_{(2)} \rightarrow 7_{(10)} $ 을 출력
- `^` : xor 연산자로 피연산자 값이 다르면 1을 출력, 같으면 0을 출력.  따라서 $ 011_{(2)} \rightarrow 3_{(10)} $ 을 출력

In [22]:
-5&15, -5|15, -5^15

(11, -1, -12)

In [26]:
bin(-5), bin(15)

('-0b101', '0b1111')

5  -> 0 0101 <br>
-5 -> 1 1011 (1의 보수를 한 후 1을 더함 == 2의 보수) <br> 
15 -> 0 1111

**설명**: 

- `&` : and 연산자로 피연산자값 값이 모두 1이여야 1이 됨, 1이 하나만 있다면 0을 출력
- `|` : or 연산자로 피연산자값이 하나만 1이여도 1을 출력, 모두 0일 때만 0을 출력
- `^` : xor 연산자로 피연산자 값이 다르면 1을 출력, 같으면 0을 출력.

<br>

- `&`연산의 결과는 0 1011. 따라서 $ 1011_{(2)} \rightarrow 11_{(10)} $ 을 출력
- `|`연산의 결과는 1 1111. 따라서 $ 1 1111_{(2)} \rightarrow  (-16 + 15) \rightarrow -1_{(10)} $ 을 출력
- `^` 연산의 결과는 0 1011. 따라서 $ 1 0100_{(2)} \rightarrow -12_{(10)}  $ 을 출력

## Markdown에서 LaTeX 수식 관련

### 예제 3. 
아래 문장을 새로운 셀에 그대로 적어 보라.

행렬은 수 또는 다항식들을 직사각형 형태로 배열한 것이다.

$$ A = \begin{bmatrix}
            3 & 5 & 7\\
            1 & 3 & 4\\
\end{bmatrix}
\begin{bmatrix} 1 & 2 \\ 2 & 3 \\ 1 & -1\end{bmatrix}
= \begin{bmatrix} 
(3)(1) + (5)(2) + (7)(1) &  (3)(2) + (5)(3) + (7)(-1) \\
(1)(1) + (3)(2) + (4)(1) &  (1)(2) + (3)(3) + (4)(-1)
\end{bmatrix}
= \begin{bmatrix} 
20 &  14 \\
11 &  7
\end{bmatrix}
$$

**여기에 위의 식을 다시 작성하라**

- `bmatrix` : 대괄호 
- `pmatrix` : 소괄호

 

$$\
    A = \begin{bmatrix} 3 & 5 & 7\\ 1 & 3 & 4\\ \end{bmatrix} \begin{bmatrix} 1 & 2\\ 2 & 3\\ 1 & -1\\ \end{bmatrix}\
    = \begin{bmatrix} (3)(1) + (5)(2) + (7)(1) & (3)(2) + (5)(3) + (7)(-1)\\ \
    (1)(1) + (3)(2) + (4)(1) & (1)(2) + (3)(3) + (4)(-1) \end{bmatrix}\
    = \begin{bmatrix} 20 & 14\\ 11 & 7 \end{bmatrix} \
   $$





In [ ]:
# **여기에 위의 식을 다시 작성하라**

# - `bmatrix` : 대괄호 
# - `pmatrix` : 소괄호

 

# $$\
#     A = \begin{bmatrix} 3 & 5 & 7\\ 1 & 3 & 4\\ \end{bmatrix} \begin{bmatrix} 1 & 2\\ 2 & 3\\ 1 & -1\\ \end{bmatrix}\
#     = \begin{bmatrix} (3)(1) + (5)(2) + (7)(1) & (3)(2) + (5)(3) + (7)(-1)\\ \
#     (1)(1) + (3)(2) + (4)(1) & (1)(2) + (3)(3) + (4)(-1) \end{bmatrix}\
#     = \begin{bmatrix} 20 & 14\\ 11 & 7 \end{bmatrix} \
#    $$





- NumPy의 다차원 배열을 이용하면 위의 행렬 곱 연산을 다음과 같이 구할 수 있다.

In [28]:
import numpy as np
a = np.array([[3,5,7],[1,3,4]])
b = np.array([[1,2],[2,3],[1,-1]])
a@b

array([[20, 14],
       [11,  7]])

`@` 는 행렬의 곱

### 예제 4. while loop의 break와 continue

다음 예제 프로그램을 실행하라

In [29]:
n = 1
while n < 7:
    print(n)
    n = n + 1
print('End of program.')

1
2
3
4
5
6
End of program.


#### While loop 안에서  break 사용

In [30]:
n = 1
while n < 7:
    if (n % 4 == 0):
        break
    print(n)
    n = n+1
print('End of program.')

1
2
3
End of program.


##### While loop 안에서  continue 사용

In [31]:
n = 1
while n < 7:
    if (n % 4 == 0):
        n = n + 1
        continue
    print(n)
    n = n+1
print('End of program.')

1
2
3
5
6
End of program.


### 예제 5. 
예제 4에 있는 세 개의 프로그램을 for loop을 이용해서 각각 다시 작성하라.

In [117]:
# n = 1
# while n < 7:
#     print(n)
#     n = n + 1
# print('End of program.')


for i in range(1,7,1):
    print(i)
    
print("End of Program")

1
2
3
4
5
6
End of Program


In [121]:
# n = 1
# while n < 7:
#     if (n % 4 == 0):
#         break
#     print(n)
#     n = n+1
# print('End of program.')

for n in range(1,7):
    if(n%4==0):
        break
    print(n)
    
print("End of program")

1
2
3
End of program


In [120]:
# n = 1
# while n < 7:
#     if (n % 4 == 0):
#         n = n + 1
#         continue
#     print(n)
#     n = n+1
# print('End of program.')

for n in range(1,7):
    if (n % 4 == 0):
        continue
    print(n)
print('End of program.')

1
2
3
5
6
End of program.


### 예제 6. 
다음 리스트 또는 튜플 안의 수의 합을 구하는 프로그램을 작성하라.

In [37]:
numbers = [1, 5, 3, 6, 8]
sum(numbers)

23

In [126]:
numbers = (1, 5, 3, 6, 8)

s1 = 0
for i in range(5):
    s1 += numbers[i]
print(s1)

23


In [127]:
numbers = [1, 5, 3, 6, 8]

s2 = 0
for i in range(len(numbers)):
    s2 += numbers[i]
    
print(s2)

23


In [128]:
numbers = (1, 5, 3, 6, 8)

s3=0
for i in numbers:
    s3 += i
    
print(s3)

23


### 예제 7. 
예제 6의 프로그램을 튜플의 모든 수를 곱한 결과를 출력하도록 수정하라.

In [53]:
numbers = (1, 5, 3, 6, 8)

mul = 1
for i in numbers:
    # print(i)
    mul *= i

print(mul)

720


### 예제 8. 
예제 6의 프로그램에서 튜플에 숫자가 아닌 객체가 있는 경우를 생각해 보자. 다음 셀의 프로그램은 실행 중에 오류가 발생한다. numbers라는 리스트의 숫자만 더해서 합을 계산하도록 수정하라.

In [2]:
numbs = (1, 5.5, 3.2, 'a', 8)

In [5]:
sum_of_num = 0

for e in numbs:    
    if(type(e) != str):  # 내가 쓴 답. if(type(e) != type('a'):
        sum_of_num += e
        
print(sum_of_num)

17.7


In [137]:
numbs = (1, 5.5, 3.2, [1,3], 8)

sum_of_num = 0

for e in numbs:    
    if(type(e) == type(1) or type(e) == float):
        sum_of_num += e
        
print(sum_of_num)

17.7


In [138]:
import numbers
isinstance(3+4.5j, numbers.Number)
isinstance(3+4.5j, complex)

True

In [82]:
isinstance(True, numbers.Number)

True

In [141]:
numbs = (1, 5.5, 3.2, 'a', 8, 3+4.5j, True)

In [142]:
sum_of_num = 0

for e in numbs:    
    if(type(e) == type(1) or type(e) == float or type(e) == complex or type(e) == bool):
        sum_of_num += e
        
print(sum_of_num)

(21.7+4.5j)


In [153]:
import numbers

sum_of_num = 0

for e in numbs:    
    if(isinstance(e, numbers.Number)):
        sum_of_num += e
        
print(sum_of_num)

(21.7+4.5j)


In [143]:
type(4) == int

True

In [147]:
isinstance('str', int)

False

In [148]:
isinstance(4, int)

True

In [150]:
isinstance(3+3j, int)

False

### 예제 10. 
다음 셀을 실행해 보라.

In [175]:
import random
random_samples = random.sample(range(1, 11), 5)
print(random_samples)

[10, 2, 8, 4, 5]


In [176]:
help(random.sample)

Help on method sample in module random:

sample(population, k, *, counts=None) method of random.Random instance
    Chooses k unique random elements from a population sequence or set.
    
    Returns a new list containing elements from the population while
    leaving the original population unchanged.  The resulting list is
    in selection order so that all sub-slices will also be valid random
    samples.  This allows raffle winners (the sample) to be partitioned
    into grand prize and second place winners (the subslices).
    
    Members of the population need not be hashable or unique.  If the
    population contains repeats, then each occurrence is a possible
    selection in the sample.
    
    Repeated elements can be specified one at a time or with the optional
    counts parameter.  For example:
    
        sample(['red', 'blue'], counts=[4, 2], k=5)
    
    is equivalent to:
    
        sample(['red', 'red', 'red', 'red', 'blue', 'blue'], k=5)
    
    To choose a sa

### 예제 11. 
다음 프로그램이 무엇을 하는 프로그램인지 이해해 보자.

In [177]:
numbs = random.sample(range(1, 11), 5)
in_numb = int(input("1과 10 사이의 정수를 입력하시오: "))

found = False #sentinel
for num in numbs:
    if num == in_numb:
        found = True
        break

if found:
    print(f"{in_numb}는(은) 리스트에 포함되어 있다.")
else:
    print(f"{in_numb}는(은) 리스트에 포함되어 있지 않다.")
print(numbers)          

1과 10 사이의 정수를 입력하시오:  100


100는(은) 리스트에 포함되어 있지 않다.
<module 'numbers' from '/Users/younghoonna/opt/anaconda3/lib/python3.9/numbers.py'>


**설명**: 1부터 11보다 작은, 즉 1부터 10까지의 수 중 5개의 랜덤 수(난수)를 생성하여 리스트에 집어 넣는다. <br>
만약 내가 고른 수가 리스트에 있다면 found = True가 되어 반복문을 탈출, 그리고 내가 입력한 수가 포함되어있다는 문장을, <br>
반복문을 다 돌았는데 False가 유지되어 있다면 포함되어 있지 않다는 명령어를 출력한다.

### 예제 12. 
예제 11을 다음과 같이 수정할 수 있다. 이 예제의 for-else 구문에서 else의 블록이 언제 실행되는지 설명하라.

In [190]:
numbers = random.sample(range(1, 11), 5)
in_num = int(input("1과 10 사이의 정수를 입력하시오: "))

for num in numbers:
    if num == in_num:
        print(f"{in_num}는 리스트에 포함되어 있다.")
        break
else:
    print(f"{in_num}는 리스트에 포함되어 있지 않다.")
print(numbers)          

1과 10 사이의 정수를 입력하시오:  10


10는 리스트에 포함되어 있다.
[2, 10, 4, 5, 8]


**설명**: sentinal flag를 사용하지 않고 바로 출력문을 출력 후 반복문을 출력한다. break를 통해 반복문을 빠져나가면 else를 실행 안함. for 문이 끝난 후에 실행 되는 것이 else 구문.

###  예제 13. 
예제 11의 프로그램을 in 연산자를 사용해서 for loop을 사용하지 않는 프로그램으로 수정하라.

In [195]:
numbers = random.sample(range(1, 11), 5)
in_num = int(input("1과 10 사이의 정수를 입력하시오: "))

if in_num in numbers:
    print(f"{in_num}는 리스트에 포함되어 있다.")
else:
    print(f"{in_num}는 리스트에 포함되어 있지 않다.")
print(numbers)          

1과 10 사이의 정수를 입력하시오:  10


10는 리스트에 포함되어 있지 않다.
[6, 9, 5, 8, 7]


### 예제 14. 
예제 12에서 다른 부분은 그대로 두고  for loop만 while loop으로 수정하라.

In [17]:
import random 
numbers = random.sample(range(1, 11), 5)
in_num = int(input("1과 10 사이의 정수를 입력하시오: "))

i=0
while i<len(numbers):
    num = numbers[i]
    i+=1
    
    if num == in_num:
        print(f"{in_num}는 리스트에 포함되어 있다.")
        break
        
    print(i)


# else:
#     print(f"{in_num}는 리스트에 포함되어 있지 않다.")
       

1과 10 사이의 정수를 입력하시오:  10


1
2
3
4
10는 리스트에 포함되어 있다.


### 예제 15. 
예제 14에서 while 부분을 whlle-else 구문을 이용해서 수정하라. 참고로 while-else 구문에서 else의 역할은 for-else 구문에서 else 역할과 같다.

In [20]:
import random 
numbers = random.sample(range(1, 11), 5)
in_num = int(input("1과 10 사이의 정수를 입력하시오: "))

i=0
while i<len(numbers):
    num = numbers[i]
    i+=1
    
    if num == in_num:
        print(f"{in_num}는 리스트에 포함되어 있다.")
        break
        
    # print(i)
    
else:
    print(f"{in_num}는 리스트에 포함되어 있지 않다.")

1과 10 사이의 정수를 입력하시오:  11


11는 리스트에 포함되어 있지 않다.


### 예제 16. 

사용자가 키보드로 입력한 문자열이 숫자로만 구성된 것이면 문자열과 True를, 아니면 문자열과 False를 출력하는 과정을 무한 반복하는 프로그램을 작성하라. 단, 사용자가 'q'를 입력하면 프로그램이 종료되어야 한다.

In [31]:
while True:
    string = input("키보드로 무언가 입력 : ")

    if string == 'q':
        break;
    
    for n in range(len(string)):
        if not(string[n] >= '0' and string[n] <= '9'):
            print("False")
            break
    else:
        print("True")

키보드로 무언가 입력 :  1


True


키보드로 무언가 입력 :  2s


False


키보드로 무언가 입력 :  s


False


키보드로 무언가 입력 :  120d


False


키보드로 무언가 입력 :  21ed9


False


키보드로 무언가 입력 :  29148828158218523910001


True


키보드로 무언가 입력 :  0078


True


키보드로 무언가 입력 :  q


### 예제 17. 소수 구하는 함수 만들기



In [32]:
numb = 52
for n in range(2, numb+1):
    n_is_a_prime = True
    for k in range(2, n//2+1):
        if n % k == 0:
            n_is_a_prime = False
            break
    if n_is_a_prime:
        print(n, end=" ")

2 3 5 7 11 13 17 19 23 29 31 37 41 43 47 

In [33]:
numb = 52

def get_primes(numb):
    # parameter : numb (int)
    # return    : primes
    
    primes = [] # 빈 리스트.
    for n in range(2, numb+1):
        n_is_a_prime = True
        for k in range(2, n//2+1):
            if n % k == 0:
                n_is_a_prime = False
                break
        if n_is_a_prime:
            primes.append(n)
            
    return primes

In [209]:
a = []
a

[]

In [210]:
a.append(5)
a

[5]

### 문제 1. 
임의의 십진수을 키보드로 입력받아서 이진수 문자열로 변환하여 콘솔에 출력하는 Python 프로그램을 작성하라. 출력 포맷은 bin() 함수를 참조하고, bin() 함수를 사용하면 안된다.

In [39]:
n = int(input("Num : "))

string = ''
while (n > 1):
    string += str(n%2)
    n //= 2
    
string += str(n%2)

print()
for i in range(len(string)-1, -1, -1):
    print(string[i], end='')
    
    # print(i)

Num :  110



1101110

In [ ]:
n = int(input("10진수 입력 :"))
print(n)
string = ""
while n > 1:
    string+=str(n%2)        
    n//=2
    

string+=str(n%2)        

reverse = "0b"
for s in range(len(string)-1, -1, -1):
    reverse+= string[s]
    
reverse

### bin() 함수와 출력문을 똑같이 적용.

In [41]:
n = int(input("Num : "))

string = ''
while (n > 1):
    string += str(n%2)
    n //= 2
    
string += str(n%2)

result = '\'0b'
for i in range(len(string)-1, -1, -1):
    result += string[i]    

result += '\''
print(result)

Num :  110


'0b1101110'


In [40]:
bin(110)

'0b1101110'

# 문제 2. Goldbach의 추측

- 2보다 큰 모든 짝수는 소수 2 개의 합으로 표현될 수 있다
- 예
  - $4 = 2 + 2$
  - $6 = 3 + 3$
  - $8 = 3 + 5$
  - $10 = 3 + 7$ 또는 $5 + 5$
  - $12 = 5 + 7$
  - $14 = 3 + 11$
  - $16 = 3 + 13$ 또는 $5 +7$
  - $18 = 5 + 13$ 또는 $7 + 11$

사용자가 2보다 큰 짝수를 입력하면 이 짝수를 두 개의 소수의 합으로 표현하는 프로그램을 작성하라. 
- 사용자의 입력이 2보다 큰 짝수가 아니면 2보다 큰 짝수를 입력하라는 메시지를 출력하고 다시 입력을 기다린다.
- 두 소수를 구하면 사용자가 입력한 정수를 두 수의 합으로 표현하는 문자열을 출력하고 다시 사용자의 입력을 기다린다.
- 사용자의 입력이 'q'이면 프로그램을 종료한다.

In [88]:
import math

while(True):
    n = input("Num : ")
    
    if(n == 'q'):
        break  
    else:
        if(int(n)%2 ==0 and int(n) > 2):
            n = int(n)
        else:
            print(" 2보다 큰 짝수가 아닙니다. 다시 입력하세요")

    num = get_primes(n)
    
    print(n, end= " = ")
    b = False

    for t in num:
        for j in range(len(num)-1, -1, -1):
            if t > n/2:
                break
                
            if b and (t+num[j] == n):
                print(" 또는 ", end = '')
            
            if (t+num[j] == n):
                b = True
                print(t, "+", num[j], end = '')
    

Num :  10


10 = 3 + 7 또는 5 + 5

Num :  4


4 = 2 + 2

Num :  6


6 = 3 + 3

Num :  8


8 = 3 + 5

Num :  10


10 = 3 + 7 또는 5 + 5

Num :  12


12 = 5 + 7

Num :  14


14 = 3 + 11 또는 7 + 7

Num :  16


16 = 3 + 13 또는 5 + 11

Num :  18


18 = 5 + 13 또는 7 + 11

Num :  20


20 = 3 + 17 또는 7 + 13

Num :  22


22 = 3 + 19 또는 5 + 17 또는 11 + 11

Num :  q


## 시간복잡도를 줄여보기 위해 약수 구하는 법으로 에라토스테네스의 체를 사용해 보았습니다.
- 그 외에도 이번에는 다르게 한번에 여러개를 입력받고 그 후에 한번에 출력하는 형식으로 도전해 보았습니다.

In [89]:
## import math

num = []
while(True):
    n = input("Num : ")
    
    if(n == 'q'):
        break  
    else:
        n = int(n)
    
    # 일단 2보다 큰 짝수라면 list로 넣기.
    if(int(n)%2 ==0 and int(n) > 2):
        num.append(int(n))
    
    else:
         print(" 2보다 큰 짝수가 아닙니다. 다시 입력하세요")

          
max_num = max(num)

print("======= 입력으로 들어온 2보다 큰 짝수 ========")
print(num)


li = []; li.append(1); li.append(1)

for i in range(2, max_num+1):
    li.append(0)

#에라토스테네스의 체. 매번 실행하는 것이 아닌 제일 큰 숫자 하나만 구하기.
for i in range(2, round(math.sqrt(max_num))+1):
    for j in range(2, max_num//i+1):
        li[i*j] += 1
        

eratos = []
#에라토스테네스의 체로 걸러낸 결과
for i in range(0, max_num+1):
    if(li[i] == 0):
        eratos.append(i)

# 출력이 잘 되었는지 확인.
print("에라토스테네스의 체로 알아낸 소수 : ", eratos)

for n in num:
    result = []
    for i in range(0, round(len(eratos)/2)+1):
        for j in eratos:

            if(eratos[i] > n/2 or j > n):
                break

            if eratos[i]+j == n:
                result.append([eratos[i], j])

    #출력문.

    print("======= 출력 ========")
    print(n, end = " = ")
    for a,b in result:    
        if( [a,b] != result[len(result)-1]):
            print(a, "+", b, end = " 또는 ")
        else:
            print(a, "+", b)

Num :  10
Num :  20
Num :  30
Num :  40
Num :  50
Num :  60
Num :  q


======= 입력으로 들어온 2보다 큰 짝수 ========
[10, 20, 30, 40, 50, 60]
에라토스테네스의 체로 알아낸 소수 :  [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59]
======= 출력 ========
10 = 3 + 7 또는 5 + 5
======= 출력 ========
20 = 3 + 17 또는 7 + 13
======= 출력 ========
30 = 7 + 23 또는 11 + 19 또는 13 + 17
======= 출력 ========
40 = 3 + 37 또는 11 + 29 또는 17 + 23
======= 출력 ========
50 = 3 + 47 또는 7 + 43 또는 13 + 37 또는 19 + 31
======= 출력 ========
60 = 7 + 53 또는 13 + 47 또는 17 + 43 또는 19 + 41 또는 23 + 37
